lib/iex/lib/iex/helpers.ex

```diff
@@ -93,19 +93,30 @@ defmodule IEx.Helpers do
 
   """
   def recompile(options \\ []) do
-    if mix_started?() do
-      project = Mix.Project.get()
-
-      if is_nil(project) or
-           project.__info__(:compile)[:source] == String.to_charlist(Path.absname("mix.exs")) do
-        do_recompile(options)
-      else
-        message = "Cannot recompile because the current working directory changed"
-        IO.puts(IEx.color(:eval_error, message))
-      end
-    else
-      IO.puts(IEx.color(:eval_error, "Mix is not running. Please start IEx with: iex -S mix"))
-      :error
+    cond do
+      not mix_started?() ->
+        IO.puts(IEx.color(:eval_error, "Mix is not running. Please start IEx with: iex -S mix"))
+        :error
+
+      Mix.installed?() ->
+        Mix.in_install_project(fn ->
+          result = do_recompile(options)
```

Comment:
This actually also returns `:noop`, even if the dependencies were recompiled.

Reply:
That's because there is no code in the root project itself. I would probably replace this by a `Mix.Task.run("deps.compile")` or similar instead. We don't gain anything by calling `do_recompile/1`.

Reply:
Don't we want the path consolidation logic?

`Mix.Task.run("deps.compile")` doesn't distinguish the status and always returns `:ok`, so it doesn't really matter. Let's just always return `:ok`?

Reply:
path consolidation or protocol consolidation? If you mean the second, then you are right. :)

Reply:
`Mix.Task.run("compile")` is also consistent with Mix.install. I changed it to just return `:ok`, but if you have other preference let me know.